Overwrite existing inner class info from JSON.

Kobata opened the pull request:
By ignoring any pre-existing inner class info attempts to edit it,
such as by ForgeGradle handing AccessTransformers on inner classes,
gets ignored.

Some inner classes still work, but many anonymous ones, and some named
inner classes such as the GuiContainerCreative ones have pre-existing
inner class info for some reason.

(This seems to be every inner class in the JSON supplied by Forge with
a "start" attribute)

Kobata committed:
Overwrite existing inner class info from JSON.

By ignoring any pre-existing inner class info attempts to edit it,
such as by ForgeGradle handing AccessTransformers on inner classes,
gets ignored.

Some inner classes still work, but many anonymous ones, and some named
inner classes such as the GuiContainerCreative ones have pre-existing
inner class info for some reason.

(This seems to be every inner class in the JSON supplied by Forge with
a "start" attribute)

LexManos commented:
1) Please follow the code standard when making PRs
2) This should never be an issue in Forge/FML, as there should never be any Inner class information when you're doing a decomp workspace. Which is when your AT's would have any effect.

What exactly are you doing that causes inner classes to exist?
This was designed to inject anything that was missing, NOT modify existing. Which is why I chose to write those flags that way.
If we were to go your route we'd have to switch from a simple boolean to a list of visited entries, and we'd have to take into account the same access level information we do in our ATs to make sure the 'most visible' one wins.

Kobata commented:
I'm not doing anything other than an AT containing the following two lines that are an issue:

```
public net.minecraft.client.gui.inventory.GuiContainerCreative$ContainerCreative
public net.minecraft.client.gui.inventory.GuiContainerCreative$CreativeSlot
```

Somehow, the deobf (pre-exception info) jar generated by plain ForgeGradle already contains an innerClasses secton for these two classes, and from log output not listing addition of inner class info, apparently many others.

The ATs apply correctly to the main class structure, but not to the innerClass information.
Fernflower doesn't seem to care, even though javac does, so they show up in the decompiled source (and thus it seems like I can somewhat bypass it by always running setupDecompile now that that recompiles it for use).

Kobata commented:
Here's a list of all classes that are logged with an 'Overwriting Inner Class' with this modification using the configurations generated by ForgeGradle while testing:
https://gist.github.com/Kobata/bd522d56a52a2629eab4

LexManos commented:
It would be all classes that we edit in Forge/FML that have inner classes, We do not strip out this information during reobf so the binpatches will see it there.
Cleanup this PR and make it respect partial lists properly, and i'll accept it. 
So:
  1) Fix Code Style
  2) Have proper 'most visible' access system like we have in the normal ATs
  3) Support for adding any missing classes. Not just chunk replace.

Kobata commented:
1) Alright, I'll have to pay closer attention to that.
2) I think this only supports single edits, since the JSON only lists each class twice, once in itself and once in its parent class. Isn't the point of [fixAccess in ForgeGradle](https://github.com/MinecraftForge/ForgeGradle/blob/master/src/main/java/net/minecraftforge/gradle/tasks/ProcessJarTask.java#L235) to do the 'most visible' processing?
3) I'm not quite sure exactly what you mean by 'missing classes' in this context.

Edit: Oh, I think I see what you mean by #2 now -- most visible between JSON and pre-existing. Yeah, that's something that probably needs to be accounted for.

LexManos commented:
2) Yes, but what if the JSON says to make it protected. But the class file itself says to make it public? The class file should win.
3) Class file has: ONLY $Bob, Json has $Bob and $Sally, resulting jar should have both $Bob and $Sally

Kobata committed:
Handle final bits as well.

In this case, unset final if either are unset, as that
is the most permissive setting.

Kobata commented:
Hopefully I didn't miss any of the code style issues this time.

I've got it merging the access levels/final bits since those are the most important ones that might mismatch.

As for handing things that don't exist in the class: I didn't remove the code that did that already, this only handles the case where a specific class is already mentioned. Anything missing is never visited in these two methods and added by the original code in visitEnd.

It seems like there's a potentially 'better' solution to be found by reading the access field from the class itself, but doing so would require a two-pass process because the inner classes can be processed after the outer class.

Kobata commented:
Any feedback on the latest version of this?
